File: src/org/stringtemplate/v4/STRawGroupDir.java
package org.stringtemplate.v4;

import org.antlr.runtime.CharStream;
import org.antlr.runtime.CommonToken;
import org.stringtemplate.v4.compiler.*;
import org.stringtemplate.v4.compiler.Compiler;
import org.stringtemplate.v4.misc.Misc;

import java.net.URL;

/** A directory of templates without headers like ST v3 had.  Still allows group
 *  files in directory though like {@link STGroupDir} parent.
 */
public class STRawGroupDir extends STGroupDir {
	public STRawGroupDir(String dirName) {
		super(dirName);
	}

	public STRawGroupDir(String dirName, char delimiterStartChar, char delimiterStopChar) {
		super(dirName, delimiterStartChar, delimiterStopChar);
	}

	public STRawGroupDir(String dirName, String encoding) {
		super(dirName, encoding);
	}

	public STRawGroupDir(String dirName, String encoding, char delimiterStartChar, char delimiterStopChar) {
		super(dirName, encoding, delimiterStartChar, delimiterStopChar);
	}

	public STRawGroupDir(URL root, String encoding, char delimiterStartChar, char delimiterStopChar) {
		super(root, encoding, delimiterStartChar, delimiterStopChar);
	}

	@Override
	public CompiledST loadTemplateFile(String prefix, String unqualifiedFileName,
									   CharStream templateStream)
	{
		String template = templateStream.substring(0, templateStream.size() - 1);
		String templateName = Misc.getFileNameNoSuffix(unqualifiedFileName);
		String fullyQualifiedTemplateName = prefix + templateName;
		CompiledST impl = new Compiler(this).compile(fullyQualifiedTemplateName, template);
		CommonToken nameT = new CommonToken(STLexer.SEMI); // Seems like a hack, best I could come up with.
		nameT.setInputStream(templateStream);
		rawDefineTemplate(fullyQualifiedTemplateName, impl, nameT);
		impl.defineImplicitlyDefinedTemplates(this);
		return impl;
	}
}

File: objc/main/STRawGroupDir.m
#import <Foundation/Foundation.h>
#import "STRawGroupDir.h"

/** A dir of templates w/o headers like ST v3 had.  Still allows group files
 *  in dir though like STGroupDir parent.
 */
@implementation STRawGroupDir
+ (STRawGroupDir *) newSTRawGroupDir:(NSString *) aDirName
{
    return [[STRawGroupDir alloc] init:aDirName encoding:NSASCIIStringEncoding delimiterStartChar:'<' delimiterStopChar:'>'];
}

+ (STRawGroupDir *) newSTRawGroupDir:(NSString *) aDirName
                  delimiterStartChar:(char) aDelimiterStartChar
                   delimiterStopChar:(char) aDelimiterStopChar
{
    return [[STRawGroupDir alloc] init:aDirName encoding:NSASCIIStringEncoding delimiterStartChar:aDelimiterStartChar delimiterStopChar:aDelimiterStopChar];
}

+ (STRawGroupDir *) newSTRawGroupDir:(NSString *) aDirName
                            encoding:(NSStringEncoding) anEncoding
{
    return [[STRawGroupDir alloc] init:aDirName encoding:anEncoding delimiterStartChar:'<' delimiterStopChar:'>'];
}

+ (STRawGroupDir *) newSTRawGroupDir:(NSString *) aDirName
                            encoding:(NSString *) anEncoding
                  delimiterStartChar:(char) aDelimiterStartChar
                   delimiterStopChar:(char) aDelimiterStopChar
{
    return [[STRawGroupDir alloc] init:aDirName encoding:NSASCIIStringEncoding delimiterStartChar:aDelimiterStartChar delimiterStopChar:aDelimiterStopChar];
}

+ (STRawGroupDir *) newSTRawGroupDirWithURL:(NSURL *) aRoot
                                   encoding:(NSStringEncoding) anEncoding
                         delimiterStartChar:(char) aDelimiterStartChar
                          delimiterStopChar:(char) aDelimiterStopChar
{
    return [[STRawGroupDir alloc] initWithURL:aRoot encoding:NSASCIIStringEncoding delimiterStartChar:aDelimiterStartChar delimiterStopChar:aDelimiterStopChar];
}

- (id) init:(NSString *)aDirName encoding:(NSStringEncoding)theEncoding delimiterStartChar:(unichar)aDelimiterStartChar delimiterStopChar:(unichar)aDelimiterStopChar
{
    self = [super init:aDirName encoding:(NSStringEncoding)theEncoding delimiterStartChar:(unichar)aDelimiterStartChar delimiterStopChar:(unichar)aDelimiterStopChar];
    if ( self != nil ) {
       
    }
    return self;
}

- (id) initWithURL:(NSURL *)aRoot
          encoding:(NSStringEncoding)theEncoding
delimiterStartChar:(unichar)aDelimiterStartChar
 delimiterStopChar:(unichar)aDelimiterStopChar
{
    self = [super initWithURL:aRoot
                     encoding:(NSStringEncoding)theEncoding
           delimiterStartChar:(unichar)aDelimiterStartChar
            delimiterStopChar:(unichar)aDelimiterStopChar];
    if ( self != nil ) {
       
    }
    return self;
}

- (CompiledST *)loadTemplateFile:(NSString *) aPrefix
                        fileName:(NSString *) unqualifiedFileName
                  templateStream:(id<CharStream>) aTemplateStream
{
    NSString *template = [aTemplateStream substringWithRange:NSMakeRange(0, [aTemplateStream size] - 1)];
    NSString *templateName = [Misc getFileNameNoSuffix:unqualifiedFileName];
    CompiledST *impl = [[Compiler newCompiler] compile:templateName template:template];
    impl.prefix = aPrefix;
    return impl;
}
@end
